DS-2877

PhilipVis opened the pull request:
This is the PR for contribution [DS-2877](https://jira.duraspace.org/browse/DS-2877).
We look forward to your feedback.

PhilipVis committed:
DS-2877

tdonohue commented:
@PhilipVis : Just a quick note, in case you haven't noticed. This PR is failing unit tests.

PhilipVis committed:
DS-2877: config property external-sources.elsevier.key disabled by default

PhilipVis commented:
Thanks for the heads up. Greatly appreciated. I've resolved the error and all should be working well now.

tdonohue commented on dspace/config/modules/external-sources/elsevier.cfg:
```diff
@@ -1,5 +1,5 @@
 # Api key to be able to make the calls to retrieve the articles, this will need to be requested by the appropriate instance
-external-sources.elsevier.key = ${external-sources.elsevier.key}
+#external-sources.elsevier.key = ${external-sources.elsevier.key}
```
This is a circular config reference. I'm not understanding why you'd even have this in a comment, as it's not useful when uncommented. This almost seems to be based on the DSpace 5.x way of managing configs, where config overrides required such odd circular references. In DSpace 6.x (with Apache Commons Config) that is no longer needed, as you can override values in `local.cfg` without these references.  This commented out line should likely have the circular reference removed, as it's not useful.  Or, you could uncomment it again and give it an empty value by default, allowing it to be overridden in `local.cfg`

tdonohue commented on dspace/config/spring/api/general-import-service.xml:
```diff
@@ -22,7 +22,7 @@
 
     <bean class="java.lang.String" id="apiKey" autowire-candidate="true">
-        <constructor-arg value="${external-sources.elsevier.key}"/>
+        <constructor-arg value=""/>
```
Why is this now an empty argument? That seems rather odd to me, as it implies to me that to enable this, a user will need to edit this Spring config?  I'd recommend against forcing users to modify Spring configs, if at all possible.

PhilipVis committed:
DS-2877: config property external-sources.elsevier.key empty by default

PhilipVis commented:
Thanks for the feedback! The `external-sources.elsevier.key` property is now uncommented with an empty value by default. This configuration property will also be used instead of the spring apiKey bean so users will not have to edit the spring configuration to add a key.

pnbecker commented on dspace-api/src/main/java/org/dspace/fileaccess/FileAccessConsumer.java:
```diff
@@ -0,0 +1,113 @@
+                }
+                break;
+            default:
+                log.warn("consume() got unrecognized event: " + event.toString());
```
I'm afraid this could spam the logs. I made bad experiences with logging when developing the DOIConsumer. Please reduce the log level to debug.

pnbecker commented on dspace-api/src/main/java/org/dspace/authorize/service/AuthorizeService.java:
```diff
@@ -449,4 +449,11 @@ public void authorizeAction(Context c, EPerson e, DSpaceObject o, int action, bo
 	void switchPoliciesAction(Context context, DSpaceObject dso, int fromAction, int toAction)
 			throws SQLException, AuthorizeException;
 
```
Could you please add JavaDoc to service methods? What should they do, what attributes do they get?

pnbecker commented on dspace-api/src/main/java/org/dspace/content/Community.java:
```diff
@@ -46,7 +46,7 @@
     private final List<Community> subCommunities = new ArrayList<>();
 
-    @ManyToMany(fetch = FetchType.LAZY, mappedBy = "subCommunities")
+    @ManyToMany(fetch = FetchType.EAGER, mappedBy = "subCommunities")
```
Is that necessary? I'm afraid this could create problems for repositories with a lot of communities.

pnbecker commented on dspace-api/src/main/java/org/dspace/fileaccess/FileAccessFromMetadataServiceImpl.java:
```diff
@@ -0,0 +1,48 @@
+import org.dspace.core.*;
+import org.dspace.fileaccess.service.*;
+import org.dspace.importer.external.elsevier.entitlement.*;
+
```
I think I have a rough idea why you want to add this class. But could you please add a class comment why such a stub is added to DSpace?

pnbecker commented on dspace-api/src/main/java/org/dspace/fileaccess/ItemMetadataServiceImpl.java:
```diff
@@ -0,0 +1,43 @@
+
+import org.dspace.content.*;
+import org.dspace.fileaccess.service.*;
+
```
Please add a comment to the class, why this sub is part of DSpace.

pnbecker commented on dspace-api/src/main/java/org/dspace/fileaccess/service/FileAccessFromMetadataService.java:
```diff
@@ -0,0 +1,33 @@
+ * Created by Philip Vissenaekens (philip at atmire dot com)
+ * Date: 26/10/15
+ * Time: 11:55
+ */
```
Please add some JavaDoc at least to the services.

pnbecker commented on dspace-api/src/main/java/org/dspace/fileaccess/service/ItemMetadataService.java:
```diff
@@ -0,0 +1,64 @@
+
+import org.dspace.content.*;
+import static org.dspace.fileaccess.factory.FileAccessServiceFactory.*;
+
```
Please add some JavaDoc at least to the services.

pnbecker commented on dspace-api/src/main/java/org/dspace/importer/external/elsevier/UpdateElsevierItems.java:
```diff
@@ -0,0 +1,297 @@
+                boolean identical = fileAccessFromMetadataService.fileAccessIdentical(context, bitstream);
+
+                if(!identical){
+                    boolean overruled = Boolean.parseBoolean(bitstreamService.getMetadata(bitstream, "workflow.fileaccess.overruled"));
```
How does this work? The way I read that code, every bitstream that does not have a boolean metadata "workflow.fileaccess.overruled" might be changed, if this command line tool runs with the option `-p` and without the option `-i`. What do I miss?

pnbecker replied:
How do we make sure that "only" Bitstreams imported from Elsevier gets changed?

pnbecker commented:
Out of curiosity I took a look on small parts of this huge PR. I added some notes above. I haven't done a full review. While I think a full review is necessary, I honestly don't know if we are able to do one.

Beside the technical review we have to discuss the features them self. @tdonohue already requested review by DCAT and/or Leedership Group & Steering committee in the ticket to this PR. I expect those aspects to be discussed in DSpace's committer group as well.

PhilipVis committed:
DS-2877

PhilipVis commented:
@pnbecker Thank you for taking the time to review this PR. Your feedback is highly appreciated! 

We have added JavaDoc to the new services and methods to explain what they should do and what attributes they should get. We also implemented the changes/fixes you suggested.

The UpdateElsevierItems script has been adapted to make sure it will only update items that have been imported from an Elsevier source. Bitstreams of items that are not imported from an Elsevier source will not be updated (unless the item handle is specified with argument -i).

tdonohue commented:
Closing PR per comments on ticket:
https://jira.duraspace.org/browse/DS-2877
